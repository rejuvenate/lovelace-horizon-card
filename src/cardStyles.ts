import { css } from 'lit'

export default css`
  :host {
    --hc-primary: var(--primary-text-color);
    --hc-secondary: var(--secondary-text-color);

    --hc-field-name-color: var(--hc-secondary);
    --hc-field-value-color: var(--hc-primary);

    --hc-day-color: #8ebeeb;
    --hc-night-color: #393b78;

    --hc-accent: #d7d7d7;
    --hc-lines: var(--hc-accent);

    --hc-sun-hue: 44;
    --hc-sun-saturation: 93%;
    --hc-sun-lightness: 67%;
    --hc-sun-hue-reduce: 0;
    --hc-sun-saturation-reduce: 0%;
    --hc-sun-lightness-reduce: 0%;
    --hc-sun-color: hsl(
      calc(var(--hc-sun-hue) - var(--hc-sun-hue-reduce)),
      calc(var(--hc-sun-saturation) - var(--hc-sun-saturation-reduce)),
      calc(var(--hc-sun-lightness) - var(--hc-sun-lightness-reduce))
    );

    --hc-moon-hue: 52;
    --hc-moon-saturation: 77%;
    --hc-moon-lightness: 57%;
    --hc-moon-saturation-reduce: 0%;
    --hc-moon-lightness-reduce: 0%;
    --hc-moon-color: hsl(
      var(--hc-moon-hue),
      calc(var(--hc-moon-saturation) - var(--hc-moon-saturation-reduce)),
      calc(var(--hc-moon-lightness) - var(--hc-moon-lightness-reduce))
    );
    --hc-moon-shadow-color: #eeeeee;
    --hc-moon-spot-color: rgba(170, 170, 170, 0.1);
  }

  :host(.horizon-card-dark) {
    --hc-accent: #464646;
    --hc-moon-saturation: 80%;
    --hc-moon-lightness: 74%;
    --hc-moon-shadow-color: #272727;
  }

  .horizon-card {
    padding: 0.5em;
    font-family: var(--primary-font-family);
  }

  .horizon-card-field-row {
    display: flex;
    justify-content: space-around;
    margin-top: 1em;
    margin-bottom: -0.3em;
  }

  .horizon-card-text-container {
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .horizon-card-field-name {
    color: var(--hc-field-name-color);
  }

  .horizon-card-field-value {
    color: var(--hc-field-value-color);
    font-size: 1.2em;
    line-height: 1.1em;
    text-align: center;
  }

  .horizon-card-field-value-moon-phase {
    font-size: inherit;
  }

  .horizon-card-field-moon-phase {
    --mdc-icon-size: 2em;
    color: var(--primary-color);
  }

  .horizon-card-field-value-secondary {
    font-size: 0.7em;
  }

  .horizon-card-sun-value:before {
    content: "☉";
    padding-right: 0.5em;
  }

  .horizon-card-moon-value:before {
    content: "☽";
    padding-right: 0.5em;
  }

  .horizon-card-header {
    display: flex;
    justify-content: space-around;
    margin-top: 1em;
    margin-bottom: -0.3em;
  }

  .horizon-card-header .horizon-card-text-container {
    font-size: 1.2em;
  }

  .horizon-card-footer {
    margin-bottom: 1em;
  }

  .horizon-card-title {
    margin: 1em 1em 1em 1em;
    font-size: 1.5em;
    color: var(--hc-primary);
  }

  .horizon-card-graph {
    margin: 1em 0.5em 1em 0.5em;
  }

  .horizon-card-graph .dawn {
    fill: var(--hc-night-color);
    stroke: var(--hc-night-color);
  }

  .horizon-card-graph .day {
    fill: var(--hc-day-color);
    stroke: var(--hc-day-color);
  }
`
