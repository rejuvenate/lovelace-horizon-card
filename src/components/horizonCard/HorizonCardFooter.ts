import { html, nothing, TemplateResult } from 'lit'

import {
  EHorizonCardI18NKeys,
  IHorizonCardConfig,
  THorizonCardData,
  THorizonCardFields,
  TMoonTimes,
  TSunTimes
} from '../../types'
import { HelperFunctions } from '../../utils/HelperFunctions'
import { I18N } from '../../utils/I18N'

export class HorizonCardFooter {
  private readonly data: THorizonCardData
  private readonly i18n: I18N
  private readonly sunTimes: TSunTimes
  private readonly moonTimes: TMoonTimes
  private readonly fields: THorizonCardFields
  private readonly azimuths
  private readonly azimuthExtraClasses: string[]
  private readonly elevations
  private readonly elevationExtraClasses: string[]

  constructor (config: IHorizonCardConfig, data: THorizonCardData, i18n: I18N) {
    this.data = data

    this.i18n = i18n
    this.sunTimes = data.sunData.times
    this.moonTimes = data.moonData.times
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    this.fields = config.fields!

    this.azimuths = []
    if (this.fields.sun_azimuth) {
      this.azimuths.push(this.data.sunData.azimuth)
    }
    if (this.fields.moon_azimuth) {
      this.azimuths.push(this.data.moonData.azimuth)
    }
    if (this.fields.sun_azimuth && this.fields.moon_azimuth) {
      this.azimuthExtraClasses = ['horizon-card-sun-value', 'horizon-card-moon-value']
    } else {
      this.azimuthExtraClasses = []
    }

    this.elevations = []
    if (this.fields.sun_elevation) {
      this.elevations.push(this.data.sunData.elevation)
    }
    if (this.fields.moon_elevation) {
      this.elevations.push(this.data.moonData.elevation)
    }
    if (this.fields.sun_elevation && this.fields.moon_elevation) {
      this.elevationExtraClasses = ['horizon-card-sun-value', 'horizon-card-moon-value']
    } else {
      this.elevationExtraClasses = []
    }
  }

  public render (): TemplateResult {
    return html`
      <div class="horizon-card-footer">
        ${
          this.renderRow(
            this.fields.dawn
              ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Dawn, this.sunTimes.dawn)
              : nothing,
            this.fields.noon
              ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Noon, this.sunTimes.noon)
              : nothing,
            this.fields.dusk
              ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Dusk, this.sunTimes.dusk)
              : nothing
          )
        }
        ${
          this.renderRow(
            this.fields.sun_azimuth || this.fields.moon_azimuth
              ? HelperFunctions.renderFieldElements(this.i18n, EHorizonCardI18NKeys.Azimuth, this.azimuths,
                this.azimuthExtraClasses)
              : nothing,
            this.fields.sun_elevation || this.fields.moon_elevation
              ? HelperFunctions.renderFieldElements(this.i18n, EHorizonCardI18NKeys.Elevation, this.elevations,
                this.elevationExtraClasses)
              : nothing
          )
        }
        ${
          this.renderRow(
            this.fields.moonrise
              ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Moonrise, this.moonTimes.moonrise)
              : nothing,
            this.fields.moon_phase
              ? HelperFunctions.renderMoonElement(this.i18n, this.data.moonData.phase, this.data.moonData.phaseRotation)
              : nothing,
            this.fields.moonset
              ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Moonset, this.moonTimes.moonset)
              : nothing
          )
        }
      </div>
    `
  }

  private renderRow (...args: (typeof nothing | TemplateResult)[]) {
    const nonEmpty = args.filter((tr) => tr !== nothing)
    return nonEmpty.length > 0
      ? html`
        <div class="horizon-card-field-row">
          ${nonEmpty}
        </div>`
      : nothing
  }
}
