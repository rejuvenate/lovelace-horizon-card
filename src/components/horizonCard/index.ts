export * from './HorizonCard'
export * from './HorizonCardContent'
export * from './HorizonCardFooter'
export * from './HorizonCardGraph'
export * from './HorizonCardHeader'
