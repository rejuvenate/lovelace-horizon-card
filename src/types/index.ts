import { LovelaceCardConfig, NumberFormat, TimeFormat } from 'custom-card-helpers'

export type THorizonCardFields = {
  sunrise?: boolean
  sunset?: boolean

  dawn?: boolean
  noon?: boolean
  dusk?: boolean

  azimuth?: boolean
  sun_azimuth?: boolean
  moon_azimuth?: boolean
  elevation?: boolean
  sun_elevation?: boolean
  moon_elevation?: boolean

  moonrise?: boolean,
  moonset?: boolean,
  moon_phase?: boolean
}

export interface IHorizonCardConfig extends LovelaceCardConfig {
  dark_mode?: boolean
  language?: string

  title?: string,
  moon?: boolean,
  time_format?: TimeFormat
  number_format?: NumberFormat

  latitude?: number
  longitude?: number
  elevation?: number
  southern_flip?: boolean
  moon_phase_rotation?: number
  now?: Date
  time_zone?: string
  refresh_period?: number
  debug_level?: number

  fields?: THorizonCardFields
}

export type TSunTimes = {
  now: Date
  dawn?: Date
  dusk?: Date
  midnight: Date
  noon: Date
  sunrise?: Date
  sunset?: Date
}

export type TSunData = {
  readonly azimuth: number
  readonly elevation: number
  readonly times: TSunTimes
  readonly hueReduce: number
  readonly saturationReduce: number
  readonly lightnessReduce: number
}

export type TSunPosition = {
  readonly x: number
  readonly y: number
  readonly horizonY: number
  readonly sunriseX: number
  readonly sunsetX: number
  readonly scaleY: number
  readonly offsetY: number
}

export type TMoonTimes = {
  readonly now: Date,
  readonly moonrise: Date,
  readonly moonset: Date
}

export type TMoonPhase = {
  state: string
  icon: string
}

export type SunCalcMoonPhase =
  'newMoon'
  | 'waxingCrescentMoon'
  | 'firstQuarterMoon'
  | 'waxingGibbousMoon'
  | 'fullMoon'
  | 'waningGibbousMoon'
  | 'thirdQuarterMoon'
  | 'waningCrescentMoon'

export type TMoonData = {
  readonly azimuth: number
  readonly elevation: number
  readonly fraction: number
  readonly phase: TMoonPhase
  readonly phaseRotation: number,
  readonly zenithAngle: number
  readonly parallacticAngle: number
  readonly times: TMoonTimes,
  readonly saturationReduce: number,
  readonly lightnessReduce: number
}

export type TMoonPosition = {
  readonly x: number
  readonly y: number
}

export enum EHorizonCardErrors {
}

export type THorizonCardData = {
  readonly partial: boolean
  readonly latitude: number
  readonly longitude: number
  readonly sunPosition: TSunPosition
  readonly sunData: TSunData
  readonly moonPosition: TMoonPosition
  readonly moonData: TMoonData
}

export enum EHorizonCardI18NKeys {
  Azimuth = 'azimuth',
  Dawn = 'dawn',
  Dusk = 'dusk',
  Elevation = 'elevation',
  Noon = 'noon',
  Sunrise = 'sunrise',
  Sunset = 'sunset',
  Moonrise = 'moonrise',
  Moonset = 'moonset'
}

export type THorizonCardI18NErrorKeys = {
  [key in EHorizonCardErrors]: string
}

export type THorizonCardI18NKeys = { [key in EHorizonCardI18NKeys ]?: string } | { errors: THorizonCardI18NErrorKeys }
