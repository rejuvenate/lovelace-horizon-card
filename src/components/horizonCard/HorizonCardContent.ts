import { html, TemplateResult } from 'lit'

import { IHorizonCardConfig,THorizonCardData } from '../../types'
import { I18N } from '../../utils/I18N'
import { HorizonCardFooter } from './HorizonCardFooter'
import { HorizonCardGraph } from './HorizonCardGraph'
import { HorizonCardHeader } from './HorizonCardHeader'

export class HorizonCardContent {
  private config: IHorizonCardConfig
  private data: THorizonCardData
  private i18n: I18N

  constructor (config: IHorizonCardConfig, data: THorizonCardData, i18n: I18N) {
    this.config = config
    this.data = data
    this.i18n = i18n
  }

  render (): TemplateResult {
    return html`
      <ha-card>
        <div class="horizon-card">
          ${ this.renderHeader() }
          ${ this.renderGraph() }
          ${ this.renderFooter() }
        </div>
      </ha-card>
    `
  }

  private renderHeader (): TemplateResult {
    return new HorizonCardHeader(this.config, this.data, this.i18n).render()
  }

  private renderGraph (): TemplateResult {
    return new HorizonCardGraph(this.config, this.data).render()
  }

  private renderFooter (): TemplateResult {
    return new HorizonCardFooter(this.config, this.data, this.i18n).render()
  }
}
