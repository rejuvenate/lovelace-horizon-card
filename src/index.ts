import './components/horizonCard'
import './components/horizonCardEditor'

