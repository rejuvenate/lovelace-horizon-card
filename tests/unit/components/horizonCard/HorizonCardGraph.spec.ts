import { HorizonCardGraph } from '../../../../src/components/horizonCard'
import { Constants } from '../../../../src/constants'
import { IHorizonCardConfig, THorizonCardData } from '../../../../src/types'
import { TemplateResultTestHelper } from '../../../helpers/TestHelpers'

describe('HorizonCardGraph', () => {
  describe('render', () => {
    it(`renders the graph with the data values when provided`, async () => {
      const config = {
        moon: true
      } as IHorizonCardConfig

      const data = {
        partial: false,
        latitude: 40.5,
        longitude: 16.7,
        sunPosition: {
          x: 50,
          y: 50,
          horizonY: 84,
          sunriseX: 200,
          sunsetX: 400,
          scaleY: 1,
          offsetY: 0,
        },
        sunData: {

        },
        moonPosition: {
          x: 100,
          y: 25
        },
        moonData: {
          azimuth: 90,
          elevation: 45,
          fraction: 1,
          phase: Constants.MOON_PHASES.fullMoon,
          zenithAngle: 0,
          times: {

          }
        }
      } as THorizonCardData

      const horizonCardGraph = new HorizonCardGraph(config, data)

      const html = await TemplateResultTestHelper.renderElement(horizonCardGraph)

      expect(html).toMatchSnapshot()
    })

    it(`renders the graph with the default values when data values are not provided`, async () => {
      const config = {
        moon: true
      } as IHorizonCardConfig

      const horizonCardGraph = new HorizonCardGraph(config, Constants.DEFAULT_CARD_DATA)

      const html = await TemplateResultTestHelper.renderElement(horizonCardGraph)

      expect(html).toMatchSnapshot()
    })

    it(`does not render the moon when not configured so`, async () => {
      const config = {
        moon: false
      } as IHorizonCardConfig

      const horizonCardGraph = new HorizonCardGraph(config, Constants.DEFAULT_CARD_DATA)

      const html = await TemplateResultTestHelper.renderElement(horizonCardGraph)

      expect(html).toMatchSnapshot()
    })
  })

  it(`renders the graph flipped horizontally when configured so`, async () => {
    const config = {
      southern_flip: true
    } as IHorizonCardConfig

    const horizonCardGraph = new HorizonCardGraph(config, Constants.DEFAULT_CARD_DATA)

    const html = await TemplateResultTestHelper.renderElement(horizonCardGraph)

    expect(html).toMatchSnapshot()
  })
})
