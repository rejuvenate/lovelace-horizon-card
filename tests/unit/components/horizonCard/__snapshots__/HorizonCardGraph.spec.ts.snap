// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizonCardGraph render does not render the moon when not configured so 1`] = `
<div class="horizon-card-graph">
  <svg viewbox="0 0 550 150"
       xmlns="http://www.w3.org/2000/svg"
  >
    <defs>
      <path id="sun-path-unscaled"
            d="M 5,146
            C 103.334,146 176.666,20 275,20
            S 446.666,146 545,146"
      >
      </path>
      <path id="sun-path"
            d="M 5,146
            C 103.334,146 176.666,20 275,20
            S 446.666,146 545,146"
      >
      </path>
      <clipPath id="upper-path-mask">
        <use href="#sun-path">
        </use>
      </clipPath>
      <clipPath id="lower-path-mask">
        <path d="M 5,146
            C 103.334,146 176.666,20 275,20
            S 446.666,146 545,146 V0 H0">
        </path>
      </clipPath>
      <filter id="moon-blur">
        <feGaussianBlur in="SourceGraphic"
                        stddeviation="0"
        >
        </feGaussianBlur>
      </filter>
      <circle id="moon"
              stroke="none"
              cx="0"
              cy="0"
              r="14"
      >
      </circle>
      <path id="shade"
            stroke-width="0"
            d="M-14,0
                a 14,14
                0
                1,1
                28,0
                a 14,14
                0
                1,0
                -28,0
                Z"
      >
      </path>
      <mask id="moon-shadow-mask">
        <use href="#shade"
             stroke="white"
             fill="white"
             filter="url(#moon-blur)"
             stroke-width="0"
        >
        </use>
      </mask>
      <mask id="moon-shadow-mask-inverted">
        <circle fill="white"
                stroke="white"
                stroke-width="0"
                cx="0"
                cy="0"
                r="14"
        >
        </circle>
        <use href="#shade"
             stroke="black"
             fill="black"
             filter="url(#moon-blur)"
             stroke-width="0"
        >
        </use>
      </mask>
    </defs>
    <g transform-origin="center"
       transform="scale(1 1)"
    >
      <line y1="3"
            y2="72"
            stroke="var(--hc-lines)"
            x1="0"
            x2="0"
      >
      </line>
      <line y1="3"
            y2="72"
            stroke="var(--hc-lines)"
            x1="0"
            x2="0"
      >
      </line>
    </g>
    <g transform-origin="center"
       transform="translate(0 0) scale(1 1)"
    >
      <use href="#sun-path"
           fill="none"
           stroke="var(--hc-lines)"
      >
      </use>
      <path clip-path="url(#lower-path-mask)"
            class="dawn"
            d="M5,0 H0 V150 H5"
      >
      </path>
      <path clip-path="url(#upper-path-mask)"
            class="day"
            d="M0,0 H0
            V0 H0"
      >
      </path>
      <line x1="5"
            x2="545"
            stroke="var(--hc-lines)"
            y1="0"
            y2="0"
      >
      </line>
      <path stroke="var(--hc-lines)"
            fill="none"
            d="M535 -5 L545 0 L535 5"
      >
      </path>
      <circle stroke="none"
              fill="var(--hc-sun-color)"
              cx="0"
              cy="0"
              r="17"
      >
      </circle>
    </g>
  </svg>
</div>
`;

exports[`HorizonCardGraph render renders the graph with the data values when provided 1`] = `
<div class="horizon-card-graph">
  <svg viewbox="0 0 550 150"
       xmlns="http://www.w3.org/2000/svg"
  >
    <defs>
      <path id="sun-path-unscaled"
            d="M 5,146
            C 103.334,146 176.666,20 275,20
            S 446.666,146 545,146"
      >
      </path>
      <path id="sun-path"
            d="M 5,146
            C 103.334,146 176.666,20 275,20
            S 446.666,146 545,146"
      >
      </path>
      <clipPath id="upper-path-mask">
        <use href="#sun-path">
        </use>
      </clipPath>
      <clipPath id="lower-path-mask">
        <path d="M 5,146
            C 103.334,146 176.666,20 275,20
            S 446.666,146 545,146 V0 H0">
        </path>
      </clipPath>
      <filter id="moon-blur">
        <feGaussianBlur in="SourceGraphic"
                        stddeviation="0"
        >
        </feGaussianBlur>
      </filter>
      <circle id="moon"
              stroke="none"
              cx="100"
              cy="25"
              r="14"
      >
      </circle>
      <path id="shade"
            stroke-width="0"
            d="M86,25
                a 14,14
                0
                1,1
                28,0
                a 14,14
                0
                1,1
                -28,0
                Z"
      >
      </path>
      <mask id="moon-shadow-mask">
        <use href="#shade"
             stroke="white"
             fill="white"
             filter="url(#moon-blur)"
             stroke-width="0"
        >
        </use>
      </mask>
      <mask id="moon-shadow-mask-inverted">
        <circle fill="white"
                stroke="white"
                stroke-width="0"
                cx="100"
                cy="25"
                r="14"
        >
        </circle>
        <use href="#shade"
             stroke="black"
             fill="black"
             filter="url(#moon-blur)"
             stroke-width="0"
        >
        </use>
      </mask>
    </defs>
    <g transform-origin="center"
       transform="scale(1 1)"
    >
      <line y1="3"
            y2="72"
            stroke="var(--hc-lines)"
            x1="200"
            x2="200"
      >
      </line>
      <line y1="3"
            y2="72"
            stroke="var(--hc-lines)"
            x1="400"
            x2="400"
      >
      </line>
    </g>
    <g transform-origin="center"
       transform="translate(0 0) scale(1 1)"
    >
      <use href="#sun-path"
           fill="none"
           stroke="var(--hc-lines)"
      >
      </use>
      <path clip-path="url(#lower-path-mask)"
            class="dawn"
            d="M5,84 H50 V150 H5"
      >
      </path>
      <path clip-path="url(#upper-path-mask)"
            class="day"
            d="M200,0 H50
            V84 H200"
      >
      </path>
      <line x1="5"
            x2="545"
            stroke="var(--hc-lines)"
            y1="84"
            y2="84"
      >
      </line>
      <path stroke="var(--hc-lines)"
            fill="none"
            d="M535 79 L545 84 L535 89"
      >
      </path>
      <circle stroke="none"
              fill="var(--hc-sun-color)"
              cx="50"
              cy="50"
              r="17"
      >
      </circle>
    </g>
    <g transform="rotate(0 100 25)">
      <use href="#moon"
           fill="var(--hc-moon-shadow-color)"
      >
      </use>
      <use href="#moon"
           fill="var(--hc-moon-color)"
           mask="url(#moon-shadow-mask)"
      >
      </use>
    </g>
    <g transform="rotate( 100 25)">
      <circle stroke="none"
              cx="96.5"
              cy="19.75"
              r="4.666666666666667"
              fill="var(--hc-moon-spot-color)"
      >
      </circle>
      <circle stroke="none"
              cx="105.25"
              cy="18"
              r="3.5"
              fill="var(--hc-moon-spot-color)"
      >
      </circle>
      <circle stroke="none"
              cx="96.5"
              cy="28.5"
              r="3.5"
              fill="var(--hc-moon-spot-color)"
      >
      </circle>
      <circle stroke="none"
              cx="107"
              cy="25"
              r="2.8"
              fill="var(--hc-moon-spot-color)"
      >
      </circle>
    </g>
  </svg>
</div>
`;

exports[`HorizonCardGraph render renders the graph with the default values when data values are not provided 1`] = `
<div class="horizon-card-graph">
  <svg viewbox="0 0 550 150"
       xmlns="http://www.w3.org/2000/svg"
  >
    <defs>
      <path id="sun-path-unscaled"
            d="M 5,146
            C 103.334,146 176.666,20 275,20
            S 446.666,146 545,146"
      >
      </path>
      <path id="sun-path"
            d="M 5,146
            C 103.334,146 176.666,20 275,20
            S 446.666,146 545,146"
      >
      </path>
      <clipPath id="upper-path-mask">
        <use href="#sun-path">
        </use>
      </clipPath>
      <clipPath id="lower-path-mask">
        <path d="M 5,146
            C 103.334,146 176.666,20 275,20
            S 446.666,146 545,146 V0 H0">
        </path>
      </clipPath>
      <filter id="moon-blur">
        <feGaussianBlur in="SourceGraphic"
                        stddeviation="0"
        >
        </feGaussianBlur>
      </filter>
      <circle id="moon"
              stroke="none"
              cx="0"
              cy="0"
              r="14"
      >
      </circle>
      <path id="shade"
            stroke-width="0"
            d="M-14,0
                a 14,14
                0
                1,1
                28,0
                a 14,14
                0
                1,0
                -28,0
                Z"
      >
      </path>
      <mask id="moon-shadow-mask">
        <use href="#shade"
             stroke="white"
             fill="white"
             filter="url(#moon-blur)"
             stroke-width="0"
        >
        </use>
      </mask>
      <mask id="moon-shadow-mask-inverted">
        <circle fill="white"
                stroke="white"
                stroke-width="0"
                cx="0"
                cy="0"
                r="14"
        >
        </circle>
        <use href="#shade"
             stroke="black"
             fill="black"
             filter="url(#moon-blur)"
             stroke-width="0"
        >
        </use>
      </mask>
    </defs>
    <g transform-origin="center"
       transform="scale(1 1)"
    >
      <line y1="3"
            y2="72"
            stroke="var(--hc-lines)"
            x1="0"
            x2="0"
      >
      </line>
      <line y1="3"
            y2="72"
            stroke="var(--hc-lines)"
            x1="0"
            x2="0"
      >
      </line>
    </g>
    <g transform-origin="center"
       transform="translate(0 0) scale(1 1)"
    >
      <use href="#sun-path"
           fill="none"
           stroke="var(--hc-lines)"
      >
      </use>
      <path clip-path="url(#lower-path-mask)"
            class="dawn"
            d="M5,0 H0 V150 H5"
      >
      </path>
      <path clip-path="url(#upper-path-mask)"
            class="day"
            d="M0,0 H0
            V0 H0"
      >
      </path>
      <line x1="5"
            x2="545"
            stroke="var(--hc-lines)"
            y1="0"
            y2="0"
      >
      </line>
      <path stroke="var(--hc-lines)"
            fill="none"
            d="M535 -5 L545 0 L535 5"
      >
      </path>
      <circle stroke="none"
              fill="var(--hc-sun-color)"
              cx="0"
              cy="0"
              r="17"
      >
      </circle>
    </g>
    <g transform="rotate(0 0 0)">
      <use href="#moon"
           fill="var(--hc-moon-shadow-color)"
      >
      </use>
      <use href="#moon"
           fill="var(--hc-moon-color)"
           mask="url(#moon-shadow-mask)"
      >
      </use>
    </g>
    <g transform="rotate(0 0 0)">
      <circle stroke="none"
              cx="-3.5"
              cy="-5.25"
              r="4.666666666666667"
              fill="var(--hc-moon-spot-color)"
      >
      </circle>
      <circle stroke="none"
              cx="5.25"
              cy="-7"
              r="3.5"
              fill="var(--hc-moon-spot-color)"
      >
      </circle>
      <circle stroke="none"
              cx="-3.5"
              cy="3.5"
              r="3.5"
              fill="var(--hc-moon-spot-color)"
      >
      </circle>
      <circle stroke="none"
              cx="7"
              cy="0"
              r="2.8"
              fill="var(--hc-moon-spot-color)"
      >
      </circle>
    </g>
  </svg>
</div>
`;

exports[`HorizonCardGraph renders the graph flipped horizontally when configured so 1`] = `
<div class="horizon-card-graph">
  <svg viewbox="0 0 550 150"
       xmlns="http://www.w3.org/2000/svg"
  >
    <defs>
      <path id="sun-path-unscaled"
            d="M 5,146
            C 103.334,146 176.666,20 275,20
            S 446.666,146 545,146"
      >
      </path>
      <path id="sun-path"
            d="M 5,146
            C 103.334,146 176.666,20 275,20
            S 446.666,146 545,146"
      >
      </path>
      <clipPath id="upper-path-mask">
        <use href="#sun-path">
        </use>
      </clipPath>
      <clipPath id="lower-path-mask">
        <path d="M 5,146
            C 103.334,146 176.666,20 275,20
            S 446.666,146 545,146 V0 H0">
        </path>
      </clipPath>
      <filter id="moon-blur">
        <feGaussianBlur in="SourceGraphic"
                        stddeviation="0"
        >
        </feGaussianBlur>
      </filter>
      <circle id="moon"
              stroke="none"
              cx="0"
              cy="0"
              r="14"
      >
      </circle>
      <path id="shade"
            stroke-width="0"
            d="M-14,0
                a 14,14
                0
                1,1
                28,0
                a 14,14
                0
                1,0
                -28,0
                Z"
      >
      </path>
      <mask id="moon-shadow-mask">
        <use href="#shade"
             stroke="white"
             fill="white"
             filter="url(#moon-blur)"
             stroke-width="0"
        >
        </use>
      </mask>
      <mask id="moon-shadow-mask-inverted">
        <circle fill="white"
                stroke="white"
                stroke-width="0"
                cx="0"
                cy="0"
                r="14"
        >
        </circle>
        <use href="#shade"
             stroke="black"
             fill="black"
             filter="url(#moon-blur)"
             stroke-width="0"
        >
        </use>
      </mask>
    </defs>
    <g transform-origin="center"
       transform="scale(-1 1)"
    >
      <line y1="3"
            y2="72"
            stroke="var(--hc-lines)"
            x1="0"
            x2="0"
      >
      </line>
      <line y1="3"
            y2="72"
            stroke="var(--hc-lines)"
            x1="0"
            x2="0"
      >
      </line>
    </g>
    <g transform-origin="center"
       transform="translate(0 0) scale(-1 1)"
    >
      <use href="#sun-path"
           fill="none"
           stroke="var(--hc-lines)"
      >
      </use>
      <path clip-path="url(#lower-path-mask)"
            class="dawn"
            d="M5,0 H0 V150 H5"
      >
      </path>
      <path clip-path="url(#upper-path-mask)"
            class="day"
            d="M0,0 H0
            V0 H0"
      >
      </path>
      <line x1="5"
            x2="545"
            stroke="var(--hc-lines)"
            y1="0"
            y2="0"
      >
      </line>
      <path stroke="var(--hc-lines)"
            fill="none"
            d="M535 -5 L545 0 L535 5"
      >
      </path>
      <circle stroke="none"
              fill="var(--hc-sun-color)"
              cx="0"
              cy="0"
              r="17"
      >
      </circle>
    </g>
  </svg>
</div>
`;
